subprojects/core/src/test/groovy/org/gradle/util/NameValidatorTest.groovy

```diff
@@ -0,0 +1,107 @@
+import org.gradle.internal.featurelifecycle.LoggingDeprecatedFeatureHandler
+import org.gradle.internal.reflect.DirectInstantiator
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.nativeplatform.internal.DefaultFlavorContainer
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
+
+@Subject(NameValidator)
+class NameValidatorTest extends Specification {
+    static forbiddenCharacters = NameValidator.FORBIDDEN_CHARACTERS
+    static forbiddenLeadingAndTrailingCharacter = NameValidator.FORBIDDEN_LEADING_AND_TRAILING_CHARACTER
+    static invalidNames = forbiddenCharacters.collect { "a${it}b"} + ["${forbiddenLeadingAndTrailingCharacter}ab", "ab${forbiddenLeadingAndTrailingCharacter}"]
+
+    @Shared
+    def domainObjectContainersWithValidation = [
+        ["artifact types", new DefaultArtifactTypeContainer(DirectInstantiator.INSTANCE, null)],
+        ["configurations", new DefaultConfigurationContainer(null, DirectInstantiator.INSTANCE, Mock(DomainObjectContext), Mock(ListenerManager), null, null, null, null, Mock(FileCollectionFactory), null, null, null, null, null, null)],
+        ["flavors",  new DefaultFlavorContainer(DirectInstantiator.INSTANCE)]
+    ]
+
+    def loggingDeprecatedFeatureHandler = Mock(LoggingDeprecatedFeatureHandler)
+
+    def setup() {
+        SingleMessageLogger.handler = loggingDeprecatedFeatureHandler
+    }
+
+    def cleanup() {
+        SingleMessageLogger.reset()
+    }
+
+    @Unroll
+    def "projects are not allowed to be named '#name'"() {
+        when:
+        new DefaultProjectDescriptor(null, name, null, Mock(ProjectDescriptorRegistry), null)
+
+        then:
+        1 * loggingDeprecatedFeatureHandler.deprecatedFeatureUsed(_  as DeprecatedFeatureUsage) >> { DeprecatedFeatureUsage usage ->
+            assertForbidden(name, usage.message)
+        }
+
+        where:
+        name << invalidNames
+    }
+
+    @Unroll
+    def "tasks are not allowed to be named '#name'"() {
+        when:
+        new TaskFactory(Mock(ClassGenerator), null, Mock(Instantiator)).create(name, DefaultTask)
+
+        then:
+        1 * loggingDeprecatedFeatureHandler.deprecatedFeatureUsed(_  as DeprecatedFeatureUsage) >> { DeprecatedFeatureUsage usage ->
+            assertForbidden(name, usage.message)
+        }
+
+        where:
+        name << invalidNames
+    }
+
+    @Unroll
+    def "#objectType are not allowed to be named '#name'"() {
+        when:
+        domainObjectContainer.create(name)
+
+        then:
+        1 * loggingDeprecatedFeatureHandler.deprecatedFeatureUsed(_  as DeprecatedFeatureUsage) >> { DeprecatedFeatureUsage usage ->
+            assertForbidden(name, usage.message)
+        }
+
+        where:
+        [name, objectType, domainObjectContainer] << [invalidNames, domainObjectContainersWithValidation].combinations().collect { [it[0], it[1][0], it[1][1]] }
+    }
+
+    private assertForbidden(name, message) {
```

Comment:
Do you have a good suggestion how to do this in this multi-dimensional test setup? I feel like just asserting that it is giving _some_ deprecation warning is ok for what we test here. If it allows us to keep the test definition compact.